wpilibNewCommands/src/main/java/edu/wpi/first/wpilibj3/command/async/AsyncScheduler.java

```diff
@@ -0,0 +1,498 @@
+  }
+
+  /**
+   * Schedules the default command for the given resource. If no default command exists, then a new
+   * idle command will be created and assigned as the default.
+   *
+   * @param resource the resource to schedule the default command for
+   */
+  public void scheduleDefault(HardwareResource resource) {
+    var defaultCommand = defaultCommands.computeIfAbsent(resource, IdleCommand::new);
+    schedule(defaultCommand);
+  }
+
+  /**
+   * Gets the currently running command that uses the given resource.
+   *
+   * @param resource the resource to query for
+   * @return the running command using a particular resource, or null if no command is currently
+   *     using it
+   */
+  public AsyncCommand getCommandUsing(HardwareResource resource) {
+    return Util.reading(lock, () -> runningCommands.get(resource));
+  }
+
+  private final ThreadLocal<AsyncCommand> commandOnCurrentThread = new ThreadLocal<>();
+
+  /**
+   * Pauses the current command for 20 milliseconds. If a higher priority command on
+   * any of the current command's required resources is scheduled while paused, and if the current
+   * command is set to {@link AsyncCommand.InterruptBehavior#Suspend}, then the command will
+   * continue to stay paused as long as any higher priority command is running. Otherwise, a higher
+   * priority command will interrupt the command while it is paused.
+   *
+   * @throws InterruptedException if the current command was canceled, or, if the command's
+   *  {@link AsyncCommand#interruptBehavior() interrupt behavior} is not set to suspend,
+   *  interrupted by a higher priority command
+   */
+  public void pauseCurrentCommand() throws InterruptedException {
+    pauseCurrentCommand(DEFAULT_UPDATE_PERIOD);
+  }
+
+  /**
+   * Pauses the current command for the specified period of time. If a higher priority command on
+   * any of the current command's required resources is scheduled while paused, and if the current
+   * command is set to {@link AsyncCommand.InterruptBehavior#Suspend}, then the command will
+   * continue to stay paused as long as any higher priority command is running. Otherwise, a higher
+   * priority command will interrupt the command while it is paused.
+   *
+   * @throws InterruptedException if the current command was canceled, or, if the command's
+   *  {@link AsyncCommand#interruptBehavior() interrupt behavior} is not set to suspend,
+   *  interrupted by a higher priority command
+   */
+  @SuppressWarnings("BusyWait")
+  public void pauseCurrentCommand(Measure<Time> time) throws InterruptedException {
+    var command = commandOnCurrentThread.get();
+    if (command == null) {
+      throw new IllegalStateException(
+          "pauseCurrentCommand() may only be called by a running command!");
+    }
+
+    long ms = (long) time.in(Milliseconds);
+    int ns = (int) (time.in(Nanoseconds) % 1e6);
+
+    // Always sleep once
+    Thread.sleep(ms, ns);
```

Comment:
Shouldn't this be waiting for 20ms - execution time so it resumes at a 20ms clock?

Reply:
Yep. Though the underlying thread jitter tends to make this moot, where the time spent parked can be off by several milliseconds. Worst I've seen is a request for 20ms actually sleep for 36... That might get improved with optimizations to reduce the number of locking calls, or maybe it just ran afoul of a stop-the-world GC pass